sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -0,0 +1,194 @@
+static kmac_error_t wait_status_bit(uint32_t bit_position, uint32_t bit_value) {
+  while (true) {
+    uint32_t reg = abs_mmio_read32(kKmacBase + KMAC_STATUS_REG_OFFSET);
+    if (bitfield_bit32_read(reg, KMAC_STATUS_ALERT_FATAL_FAULT_BIT) ||
+        bitfield_bit32_read(reg, KMAC_STATUS_ALERT_RECOV_CTRL_UPDATE_ERR_BIT)) {
+      return kKmacInternalError;
+    }
+    if (bitfield_bit32_read(reg, bit_position) == bit_value) {
+      return kKmacOk;
+    }
+  }
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_init(kmac_operation_t operation) {
+  kmac_error_t err;
+  uint32_t cfg_reg;
+  uint32_t common_cfg_reg = KMAC_CFG_SHADOWED_REG_RESVAL;
+
+  // This probably needs to be set once per device boot rather than setting it 
+  // at every KMAC invocation.
+  common_cfg_reg = bitfield_field32_write(common_cfg_reg, KMAC_CFG_SHADOWED_ENTROPY_MODE_FIELD, kEntropyEDN);
+
+  err = wait_status_bit(KMAC_STATUS_SHA3_IDLE_BIT, 1);
+  if (err != kKmacOk) {
+    return err;
+  }
+  
+  cfg_reg = common_cfg_reg;
+  switch (operation) {
+    case kSHAKE128:
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD, kKeccakStrength128);
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, kHashingModeSHAKE);
+      break;
+
+    case kSHAKE256:
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD, kKeccakStrength256);
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, kHashingModeSHAKE);
+      break;
+
+    case kSHA3_224:
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD, kKeccakStrength224);
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, kHashingModeSHA3);
+      break;
+
+    case kSHA3_256:
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD, kKeccakStrength256);
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, kHashingModeSHA3);
+      break;
+
+    case kSHA3_384:
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD, kKeccakStrength384);
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, kHashingModeSHA3);
+      break;
+
+    case kSHA3_512:
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_KSTRENGTH_FIELD, kKeccakStrength512);
+      cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, kHashingModeSHA3);
+      break;
+
+    default:
+      return kKmacNotImplemented;
+  }
+  abs_mmio_write32_shadowed(kKmacBase + KMAC_CFG_SHADOWED_REG_OFFSET, cfg_reg);
+  return kKmacOk;
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_update(const uint8_t *data, size_t data_len, uint8_t *digest, size_t digest_len) {
+  uint32_t cmd_reg = KMAC_CMD_REG_RESVAL;
+  uint32_t status_reg;
+  kmac_error_t err;
+
+  // Should we check if `data_len` > 0 and `digest_len` > 0 ?
+  // Assumption: data % 4 = 0, i.e. the data ptr is 32-bit aligned
```

Comment:
For now it's OK to just document this assumption.

Reply:
Ok, sounds good.